### ISSUE TITLE
[Suggestion] Better components inheritance

### ISSUE THREAD

**Issue by stsrki** [aspnet/Blazor#1736]
First let me explain my problem.

I have a `Field `component inside of _MyComponents_ library. All it contains is just a basic div. For example:
`Field.cshtml` in _MyComponents_
```
@inherits Base.BaseField
<div class="@ClassMapper.Class">
    @ChildContent
</div>
```
Everything good so far, it renders, and it's all good.

BUT, here is the problem, if I want to extend the Field component in another components library, let's say that this time I want it to contain a `Label`. I cannot do that. If I wan't custom logic in this new library then I must create a new `Field.cshtml` and write everything, than I have to delete `Field` from by base _MyComponents_ library because it will mess with the new component.

This is an implementation of `Field.cshtml` in _MyComponents.Bootstrap_ library:
```
@inherits BootstrapBase.BaseBootstrapField
<div class="@ClassMapper.Class">
    @if ( Label != null )
    {
        <FieldLabel>@Label</FieldLabel>
    }
    @ChildContent
</div>
```
It would be great If for example I could just inherit Field automatically, maybe if both libraries share the same component names that they would combine.

I propose two way how to do it. Maybe :)

First, here is the main `Field.cshtml` in _MyComponents_
```
@inherits Base.BaseField
<div class="@ClassMapper.Class">
    @ChildContent
</div>
```

**Suggestion 1**: do it automatically based on the component name
`Field.cshtml` in _MyComponents.Bootstrap_
```
@if ( Label != null )
{
    <FormLabel>@Label</FormLabel>
}
@ChildContent
```

**Suggestion 2**: explicitly say that it is inherited from `Field.cshtml` in _MyComponents.Bootstrap_
```
@inherits MyComponents.Field
@if ( Label != null )
{
    <FormLabel>@Label</FormLabel>
}
@ChildContent
```

**Comment by danroth27** [comment]
This sounds similar to the question raised in #1729.

Instead of using inheritance, why not use composition? I.e. have your Bootstrap Field component use the base Field component?

Also, can you clarify what you mean when you say that the parent & child components (or components with the same name) would combine? How would they combine exactly?

**Comment by stsrki** [comment]
I'm not sure what you mean by composition.

Anyway I will try to explain, and if needed maybe I can prepare you a sample project to better illustrate it.

First a project structure

> MyComponents
> Field.cshtml

> MyComponents.Bootstrap
> Field.cshtml

> MyComponents.Bulma
> Field.cshtml

So basically MyComponents.Field has just a basic structure, `<div>` and `ChildContent`
```
@inherits Base.BaseField
<div class="@ClassMapper.Class">
    @ChildContent
</div>
```
and it is inherited from an abstract class `BaseField`. Nothing fancy there.

SO, moving on to the _Bootstrap_ and _Bulma_ implementations. Because a user can choose what framework to use, I must have the same component name for all of frameworks. ie. `Field`.

And that is the problem, because when I create `Field.cshtml` in `MyComponents.Bootstrap` and `MyComponents.Bulma` project I also must remove the `Field.cshtml` from `MyComponents`. Otherwise Bootstrap and Bulma will not render.

**Bootstrap Field.cshtml**
```
@inherits BootstrapBase.BaseBootstrapField // inherited from BaseField 
<div class="some boostrap class">
    <span>Maybe a span here</span>
    <div class="@ClassMapper.Class">
        @ChildContent
    </div>
</div>
```
**Bulma Field.cshtml**
```
@inherits BulmaBase.BaseBootstrapField // inherited from BaseField 
<p class="some bulma class">    
    <div class="@ClassMapper.Class">
        @ChildContent
    </div>
    <div>Maybe a div here</div>
</p>
```
As you can see the `<div class="@ClassMapper.Class">@ChildContent</div>` is the same in both cases.

OK, moving on. Lets say that now I must implement a third framework. And it have entirely different structure.
Material Field.cshtml
```
@inherits MaterialBase.BaseMaterialField // inherited from BaseField 
<span>Hey look now only I am here</span> 
```

There can be many different structures, based on the same root component. And they all must share the same name, otherwise it will not make sence when using the library if I need to have different name for every implementation, ie. BootstrapField, BulmaField... I want to avoid that.

**Comment by stsrki** [comment]
PS. If none of this make sense or it's impossible to do, maybe there is an easier way.
Allow the `Field` to exist inside of MyComponent project and inside of any other subproject. But than the Blazor will need to somehow know which one to use.

**Comment by DeanRoddey** [comment]
Yeh, this is the same issue that I have been complaining about in 1729. I think it will end up being a ubiquitous issue if not addressed.

**Comment by wanton7** [comment]
@stsrki Could you always use `Field.cshtml` from `MyComponents`and make it as dynamic component https://github.com/aspnet/Blazor/issues/723 and have it to select actual implementation `BulmaField` etc?

**Comment by wanton7** [comment]
@DeanRoddey you can just copy & paste your issue's address from browser's address bar to your comment to link it like this https://github.com/aspnet/Blazor/issues/1729

**Comment by wanton7** [comment]
@stsrki it's up to you how create `Type` of component it doesn't have to be in the same project/assembly. `OpenComponent` uses `Type` and you can create that `Type` how ever you want.

**Comment by stsrki** [comment]
Biggest problem with your approach is that everything would have to be done programmatically. I wan't to use razor, not do everything in code. 
Lets say that I go with your route and do it programmatically. Consider this scenario:
In base project I have 
```
public abstract class BaseField
{
    protected virtual RenderFragment CreateDynamicComponent() => builder =>
    {
       // override this in implementation project
    };
}
```
then in `Field.cshtml`
```
@inherits BaseField
@CreateDynamicComponent()
```

Now in my Bulma project I override BaseField
```
public abstract class BaseBulmaField : BaseField
{
    protected override RenderFragment CreateDynamicComponent()
    {
        builder.OpenComponent(...);
        ...
        return base.CreateDynamicComponent();
    }
}
```

Now, how and where do I say to Blazor what implementation it will use? How I tell it to use BaseBulmaField?

Also actual html tree could be fairly complex. Doing it all programmatically would be a huge overhead.

**Comment by wanton7** [comment]
@stsrki It's not ideal but I didn't mean you would do it like that. I tought you could create Field component as proxy component that proxies parameters to actual implementation. So basically that would be only the interface. So your basic implementation wouldn't be in `Field.cstml` but in something like `BasicField.csthml`.

**Comment by stsrki** [comment]
Yeah but I would still be stuck with the problem that the users could not use Field component like
```
<Field>...</Field>
```
Instead they would have to use `<BulmaField>...</BulmaField>` or `<BootstrapField>...</BootstrapField>` or `<MaterialField>...</MaterialField>` or whatever. It's not a solution. The user must use only the `Field` tag after importing the right implementation from NuGet.

**Comment by wanton7** [comment]
Here is a simplified example of what I mean

*Field.cshtml*
```
@CreateDynamicComponent()

@functions {
    [Parameter]
    string Text { get; set; }

    private static int _count;

    RenderFragment CreateDynamicComponent() => builder =>
    {
        var count = System.Threading.Interlocked.Increment(ref _count);
        builder.OpenComponent(0, (count & 1) == 0 ? typeof(BasicField) : typeof(BootstrapField));
        builder.AddAttribute(1, nameof(Text), Text);
        builder.CloseComponent();
    };
}
```

*BasicField.cshtml*
```
<span>@Text</span>

@functions {
    [Parameter]
    string Text { get; set; }
}
```

*BootstrapField.cshtml*
```
<span class="label label-default">@Text</span>

@functions {
    [Parameter]
    string Text { get; set; }
}
```

**Comment by stsrki** [comment]
Yes, that's what I'm already trying to do now. But I'm stuck with the ChildContent.

_Custom.cshtml_
```
@inherits Base.BaseCustom
@if ( Type == null )
{
    <span>
        @Text
        @ChildContent
    </span>
}
else
{
    @CreateDynamicComponent()
}
```

_BaseCustom.cs_
```
public abstract class BaseCustom : BaseComponent
{
    #region Methods

    protected RenderFragment CreateDynamicComponent() => builder =>
    {
        builder.OpenComponent( 0, Type );
        builder.AddAttribute( 1, nameof( Text ), Text );
        builder.AddContent( 2, ChildContent );
        builder.CloseComponent();
    };

    #endregion

    #region Properties

    [Parameter] protected RenderFragment ChildContent { get; set; }

    [Parameter] protected string Text { get; set; }

    // This is just for testing. I will move this to a "registered components" proxy
    [Parameter] protected Type Type { get; set; }

    #endregion
}
```

_CustomBS.cshtml_ in Bootstrap project
```
@inherits Base.BaseCustom
<span class="label label-default">
    @Text
    @ChildContent
</span>
```

Test code
```
<Custom Text="AA">
    BB
</Custom>
<Custom Text="CC" Type="@(typeof(Blazorise.Bootstrap.CustomBS))">
    DD
</Custom>
```

Finally it renders
```
<span>AA BB</span>
<span class="label label-default">CC</span>
```

 It's not rendering DD value. I cannot make it to render `ChildContent` for `CustomBS` component.

**Comment by wanton7** [comment]
Something's different in your code compared to mine, tried it and worked just fine.

*Field.cshtml*
```
@CreateDynamicComponent()

@functions {
    [Parameter]
    private RenderFragment ChildContent { get; set; }

    private static int _count;

    RenderFragment CreateDynamicComponent() => builder =>
    {
        var count = System.Threading.Interlocked.Increment(ref _count);
        builder.OpenComponent(0, (count & 1) == 0 ? typeof(BasicField) : typeof(BootstrapField));
        builder.AddAttribute(1, nameof(ChildContent), ChildContent);
        builder.CloseComponent();
    };
}
```

*BasicField.cshtml*
```
<span>@ChildContent</span>

@functions {
    [Parameter]
    private RenderFragment ChildContent { get; set; }
}
```

*BootstrapField.cshtml*
```
<span class="label label-default">@ChildContent</span>

@functions {
    [Parameter]
    private RenderFragment ChildContent { get; set; }
}
```

**Comment by stsrki** [comment]
Just wanted to let you know it should have being

`builder.AddAttribute( 2, nameof( ChildContent ), ChildContent );`

You were faster to comment :)

All in all it's not an ideal solution but it works for now. Now I must see how it will behave with `onclick/onchange` events, and `bind-*` attributes.

**Comment by wanton7** [comment]
It seems that you don't have to use .cshtml you can use this instead
*Field.cs*
```
public class Field : BlazorComponent
    {
        private static int _count;

        [Parameter]
        private RenderFragment ChildContent { get; set; }

        protected override void BuildRenderTree(RenderTreeBuilder builder)
        {
            var count = System.Threading.Interlocked.Increment(ref _count);
            builder.OpenComponent(0, (count & 1) == 0 ? typeof(BasicField) : typeof(BootstrapField));
            builder.AddAttribute(1, nameof(ChildContent), ChildContent);
            builder.CloseComponent();
        }
    }
```

**Comment by stsrki** [comment]
I can but I still want for a .cshtml to have a basic html until there is a need to implement custom markup for an implementation. eg.
```
@inherits Base.BaseCustom
@if ( Type == null )
{
    <span>
        @Text
        @ChildContent
    </span>
}
else
{
    @CreateDynamicComponent()
}
```

PS. I tried the` bind-*` and a few events on inputs, and it seems good so far

**Comment by wanton7** [comment]
Here is a reusable proxy component base class. Just override ChildComponentType and be done.

```csharp
public abstract class ProxyComponent : BlazorComponent
{
    private ParameterCollection _parameters;

    public override void SetParameters(ParameterCollection parameters)
    {
        _parameters = parameters;
        base.SetParameters(ParameterCollection.Empty);
    }

    protected override void BuildRenderTree(RenderTreeBuilder builder)
    {
        var seq = 0;
        builder.OpenComponent(seq++, ChildComponentType());
        foreach (var parameter in _parameters)
            builder.AddAttribute(seq++, parameter.Name, parameter.Value);

        builder.CloseComponent();
    }

    protected abstract Type ChildComponentType();
}
```
